### ISSUE TITLE
[Req] Compatibility with Nvidia Shield

### ISSUE THREAD

**Issue by PietroPizzi69** [wolpi/prim-ftpd#257]
Hi,

your App is great but unfortunately the Google App Store says I can't install it on my Nvidia Shield.

Is this because you have to "whitelist" the Shield or because it really is not compatible?

I would love to use it on my Shield!!

**Comment by wolpi** [comment]
So far I have no idea how an app could be made visible for shield

**Comment by PietroPizzi69** [comment]
I'm no Android DEV but maybe this helps!?
https://youtu.be/Gid6c_-80EM

And for comparison, it is possible to install SimpleSSHD on my Shield:
https://play.google.com/store/apps/details?id=org.galexander.sshd
It is implemented since v23 (2019-09-21):
http://www.galexander.org/software/simplesshd/
This App is OpenSource, so you can check it out:
http://galexander.org/git/simplesshd.git
Or kindly ask greg@galexander.org about a hint.

Hope this helps!

**Comment by wolpi** [comment]
I'm going to have a look at these links ...

**Comment by wolpi** [comment]
To publish for nvidia shield on google play means to publish for android tv. An app must provide additional screenshots and a banner for that. And google has an own policiy for andriod tv. Apps which like to publish are reviewed according to this policy. This app is under review now.

**Comment by PietroPizzi69** [comment]
Nice :)! I obviously hope that it succeeds. Great that you looked into this!! If it goes through the process I will try it as soon as I have the time :)!

**Comment by Pentaphon** [comment]
> This app is under review now.

You're going to make lots of Android TV users very happy with this change. People have been looking for an open source app that opens a simple server to  easily transfer files to and from their TV device.

**Comment by wolpi** [comment]
As you can see by linked commit, there are some more changes to address Google's feedback. Not sure if they are happy with that.

**Comment by wolpi** [comment]
With Version 6.16 Google has same review results as before. Both issues are already adressed. It is unclear what Google checks.



![grafik](https://user-images.githubusercontent.com/1812560/230870911-56f8b770-bdd6-4870-891a-33f2ff01ca85.png)

**Comment by wolpi** [comment]
The DPad Issue might be solved with introduction of Tabs as in #281

**Comment by Pentaphon** [comment]
> The DPad Issue might be solved with introduction of Tabs as in #281

Thank you. A lot of people will find this useful. Have you created a banner (basically a tv-shaped icon) for the Android TV launcher like the ones you see on the screenshot of the Android TV Play Store below?

![play](https://user-images.githubusercontent.com/49374058/230872970-f4df81b2-862c-43d2-85a4-c805a33bf904.jpg)

**Comment by wolpi** [comment]
Yes, such a banner is present, as can be seen in linked commits. Obviiously Google is not happy with it :sob:

**Comment by mylove90** [comment]
Thanks for all your hard efforts....i have an old tv with Kitkat 4.4...when i open the app on it there is no buttons...will that be resolved with the release of the tv app ?

**Comment by wolpi** [comment]
Android TV uses another launcher Activity configured with another category in Manifest. It's called Leanback. Your android version obviously does not use the Leanback-Activity but the other Activity intended for phones and tablets. You can search for ways to choose another Activity when launching the app (and then choose leanback).


>will that be resolved with the release of the tv app

No, that will just make the app availble on Google Play for Android TV.

**Comment by wolpi** [comment]
Added Milestone 7.0 as changes in that version affect this issue, let's see how google review sees that.

**Comment by wolpi** [comment]
New review results are present.

Funny that they don't complain about banner anymore but still about DPad.
There has been no change regarding banner but with new tab-ui DPad has changed heavily.
This review result is completley unexpected.


![grafik](https://github.com/wolpi/prim-ftpd/assets/1812560/c5c887a7-5dd0-4435-bb73-352ebe2ab83b)

**Comment by Pentaphon** [comment]
Anything you can change on the UI to make it more D-Pad friendly?

**Comment by wolpi** [comment]
There is D-Pad support. No idea why Google rejects this.

**Comment by Pentaphon** [comment]
> There is D-Pad support. No idea why Google rejects this.

I just sideloaded primitiveFTPd-7.0-googleplay.apk on my Google TV device and while it does switch tabs when you select and press them with the remote's D-Pad, when I go to the "start server" button, it doesn't seem to actually press when using the remote. There's no way to start the server. I just keep pressing "start server" and the server doesn't start. Perhaps they found that this functionality is missing. Have you tested the app on an Android/Google TV device yourself with a remote?

**Comment by wolpi** [comment]
I'm testing it in simulator with keyboard. It works there. Good to know that there are issues with real hardware. I'll have a look. As soon as I find time.

**Comment by wolpi** [comment]
Yes, that are required changes for this topic. You can wait until I find time for them or provide a PR :wink:

**Comment by wolpi** [comment]
According to docs D-Pad should behave the same as keyboard:
* https://developer.android.com/develop/ui/views/touch-and-input/keyboard-input/navigation
* https://developer.android.com/training/tv/start/navigation.html#d-pad-navigation

So far I've been happy with default navigation order. But yes, that behaves weird in some cases. Therefore i did some adjustments.

Next issue is, that in TV mode, some elements don't show that they have focus. Thus you cannot see where your cursor is.

**Comment by wolpi** [comment]
I think with latest changes it is worth another try.

**Comment by Pentaphon** [comment]
Thank you for keeping up with this Wolpi, TV devices really benefit the most from an open source FTP server so we can send files to them easily.

**Comment by wolpi** [comment]
Version 7.1 has again been rejected. This time for another reason. This version contained a change to improve situation with copy dialog (receive shared files) and screen orientation changes. It used a simple approach to just declare a specific orientation in manifest. That is forbidden for Android TV. Thus it had to be changed, see commit above.

**Comment by PietroPizzi69** [comment]
Seams like to be a hard time to be an Android TV developer :(.

**Comment by wolpi** [comment]
Google seems to have accepted it in meantime, without sending notification.
Can you confirm?

**Comment by Leone25** [comment]
Hello,
I just tried to use this on my new TLC Android TV to allow uploading files to an external HDD I've plugged in USB.
The app starts but whenever I select `Android Access Framework (SAF)` it says "you don't have a app available for that", what I'm guessing it's happening is that my TV has no builtin file selector app. Could something be done to get around that?
Thx
Rico

**Comment by Leone25** [comment]
looks like my tv prevents me from accessing that folder, is there a way to save it somewhere else?

![Image](https://github.com/user-attachments/assets/66a5ccf3-51d7-4f16-9bd0-33a172415661)
